### DISCUSSION TITLE
Bruno - Roadmap  🛣️

### DISCUSSION THREAD

**Discussion by helloanoop** [usebruno/bruno#384] [Announcements]
Current Version: `v1.2.0`

This is a living document and will be updated and reprioritized regularly
Please comment if there are blockers that are stopping you from implementing Bruno in your teams. This will help us prioritise our roadmap accordingly.

You can also checkout our [milestones](https://github.com/usebruno/bruno/milestones), and our [changelog](https://github.com/usebruno/bruno/releases).

## Version 1 
- [x] Request and Responses
- [x] Powerful and extensive Scripting
- [x] GraphQL, SPARQL
- [x] Variable Viewer
- [x] Collection Runner  
- [x] Environments
- [x] Secret Management
- [x] Code Generation 
- [x] Asserts and Tests
- [x] Collection Level Proxy (http, https, socks5)
- [x] Client Certificates
- [x] CLI
- Package Distribution
  - [x] Binaries and Portable Versions for Mac, Linux and Win 
  - [x] Homebrew
  - [ ] Rpm
  - [x] Snap
  - [x] Apt
  - [x] Choco  
- [Auth Support](https://github.com/usebruno/bruno/issues/119)
  - [x] Bearer, Basic, AWS V4
-  [x] Request Docs
- [x] Json Prettify in Request Body
- Collection Level
  - [x] Headers, Auth, Scripts, Tests and Docs
- [x] [Open API 3.0 Imports. Beta](https://github.com/usebruno/bruno/issues/81)

## Version 2
- Package Distribution
  - [ ] Flatpak
  - [ ] Winget
  - [ ] Scoop 
- [x] [Open API 2.0 Imports](https://github.com/usebruno/bruno/issues/81)
- [Auth Support](https://github.com/usebruno/bruno/issues/119)
  - [ ] OAuth2
  - [x] Cookies
- [ ] [Save/restore workspace on close/open app](https://github.com/usebruno/bruno/issues/565)
- [ ] Improve Website Docs (Use something more powerful like mkdocs)
- [ ] [Auto Updater](https://github.com/usebruno/bruno/issues/103)
- [ ] Request Response History
- [x] Copy Curl directly into Bruno
- [Auth Support](https://github.com/usebruno/bruno/issues/119)
  - [ ] OAuth1
- Folder Level 
  - [ ] Headers, Auth, Scripts and Tests.
  - [ ] Vars, 
  - [ ] Docs 
- VSCode Extension Improvements
  - [ ] Syntax highlighting and basic support already exist. see [demo](https://www.youtube.com/watch?v=yVF13_9S6ng&t=70s)
  - [ ] Dedicate more efforts towards improving vscode extension 
- [ ] Save Response Examples
- [ ] [Nested Environments](https://github.com/usebruno/bruno/issues/292)
- [ ] [Refactor Common CLI and GUI code](https://github.com/usebruno/bruno/issues/226)
- [ ] Auto Save
- [ ] Succession Plan (In case I go dark)
- [ ] Request Runtime Module Refactoring

### **Golden Edition (Paid Edition)**
Majority of the features in Bruno (which covers REST and GraphQL Apis) are free and opensource.
We strive to strike a harmonious balance between open-source principles and sustainability - https://github.com/usebruno/bruno/discussions/269
  - [ ] Inbuilt Bru File Explorer
  - [ ] Visual Git (Like Gitlens for Vscode)
  - [ ] Inbuilt Terminal
  - [ ] GRPC, Websocket, SocketIO, MQTT
  - [ ] Load Data from File for Collection Run
  - [ ] Performance/Load Testing
  - [ ] Custom Themes
  - [ ] Vault Integration (AWS, Azure, GCP, HashiCorp)

## Version 3
- [ ] [GraphiQL explorer](https://github.com/OneGraph/graphiql-explorer)
- [ ] Web Playground (Support for hosting collections - similar to SwaggerUI)
- [ ] Collection Network (git based discovery)

## Version 4
- [ ] Mock Server
- [ ] Generate API documentation from your collections

## PROFIT 🤑
Yes, I plan to monetize my efforts as the project's creator and lead maintainer by offering a `Golden Edition` with exclusive features.
The Golden Edition will be launched soon after the v1 release.
Checkout the public discussion on [sustainable opensource](https://github.com/usebruno/bruno/discussions/269)

## Long Term Projects in Bruno's Ecosystem 
- Designer - OpenAPI designer for your APIs
- Graph - A web portal to explore, visualise and document your GraphQL APIs
- Watchdog - API Monitoring and Alerts

**Comment by nrekretep** [comment]
Does the Goal "Package Distribution" in Version 1 mean that the respective artifacts are created through the build process or will the packages also be distributed via a repository service so that one can just run `apt install bruno` after adding the bruno repo to the apt sources?

**Reply by helloanoop** [reply]
One should be able to run `apt install bruno` after adding the bruno repo to the apt sources?

**Reply by nrekretep** [reply]
Currently I have to download the `.deb` file from the Releases Page and manually install the software from the downloaded file. It would be useful to provide a debian repository server so that I can install and update `bruno` only by using `apt` (the debian package manager). 

Same should be true for other linux package formats (see also #339 )

**Reply by helloanoop** [reply]
@nrekretep Checkout https://www.usebruno.com/downloads You are going to be happy !

**Reply by nrekretep** [reply]
I am almost happy 🤪

Installing the debian package directly from a repository server via the apt tool would be very nice and would simplify application distribution in an enterprise context. 

Do you have any kind of webserver available to statically serve the debian packages? Github Pages should work also, but I did not try it yet.

**Reply by helloanoop** [reply]
@nrekretep My bad. I forgot to the add the debian repo server. Have updated it in the website now.
```bash
gpg --keyserver keyserver.ubuntu.com --recv-keys 9FA6017ECABE0266 
gpg --export 9FA6017ECABE0266 | sudo tee /etc/apt/trusted.gpg.d/bruno.gpg > /dev/null 
 
echo "deb http://debian.usebruno.com/ bruno stable" | sudo tee /etc/apt/sources.list.d/bruno.list
apt install bruno
```

**Reply by nrekretep** [reply]
Just wow! Thank you!

**Comment by ravi-bhu** [comment]
Any support for certificates ?

**Reply by rmueller83** [reply]
See https://github.com/usebruno/bruno/pull/275

**Reply by helloanoop** [reply]
Will be releasing this over the weekend! Great work done by @phoval on this ❤️

**Comment by helloanoop** [comment]
### This week's Priority List 

Mon, October 9, 2023 - Sun, October 15, 2023.

- [ ] OAuth2 Support
- [ ] Package Distribution - Homebrew, Flatpak, Rpm and Deb package repositories
- [ ] Smoother imports and graceful handling of special chars in filenames - https://github.com/usebruno/bruno/pull/349
- [ ] Mutual TLS Support - https://github.com/usebruno/bruno/pull/275

If you are implementing Bruno in your teams, and if there is a key functionality that is stopping you, please call it out in the replies

PS: The GUI file upload will take some more time, Currently file upload is only supported via scripting - https://github.com/usebruno/bruno/discussions/385#discussioncomment-7196674

**Reply by niraj-khatiwada** [reply]
Is Request docs support available?

**Reply by helloanoop** [reply]
Surprise !! Checkout Bruno v0.24.0

![image](https://github.com/usebruno/bruno/assets/9350733/52042582-8a6d-45d8-801b-dc1503e6ba2a)

**Comment by lucasfsousa** [comment]
Any support for adding the history of the requests?

**Reply by helloanoop** [reply]
Added to V1 in Roadmap

**Reply by azimhawaldar** [reply]
V2 Roadmap. Right @helloanoop  ?

**Reply by helloanoop** [reply]
Yes V2, since we decided to do a v1 release on Oct 29 2023 at IndiaFOSS 3.0

**Comment by trin4ik** [comment]
@helloanoop wow, great work. I need to document the api. need types, default values, samples, etc for all params. somethink like this:

```
post {
  url: https://api.textlocal.in/send
}

body {
  {
    "apiKey": "secret",
    "numbers": "9988776655",
    "message": "Woof! lets play with some apis",
  }
}

headers {
  content-type: application/json
  Authorization: Bearer topsecret
}

param:apiKey  {
  title: API key
  description: You can get your api key in "profile -> api keys"
  required: true
  type: string:md5
}

param:numbers {
  title: Phone number
  type: integer
  sample: [
    1234567,
    2121212
  ]
}

param:message {
  type: string
}
```

any discussion/ideas about documentation rfc in bru?

**Reply by helloanoop** [reply]
Hi @trin4ik Would https://github.com/usebruno/bruno/discussions/384#discussioncomment-7269756 solve your use case ?

**Reply by trin4ik** [reply]
@helloanoop partially. I'll try to put my suggestion in a separate discussion.

**Reply by trin4ik** [reply]
@helloanoop https://github.com/usebruno/bruno/discussions/567 that's what I mean. And that's what I often do in my team. Bru can stay lightweight as it is now, but can expand for more complex tasks I think.

**Reply by abbas-dzinehub** [reply]
How do we import the documentation we have created in postman, into bruno? The collections are getting imported fine... but the documentation isn't.

**Comment by bwgjoseph** [comment]
Would you consider distribution on Winget for windows?

**Reply by helloanoop** [reply]
This weekend, it will be done.

**Reply by cfreear** [reply]
@helloanoop any update on this?

**Reply by cfreear** [reply]
My winget is now picking up on a 1.2.0 to 1.3.2 upgrade available 🎉

**Comment by kevinrenskers** [comment]
I'm not seeing cookie support on the roadmap. Is this not planned?

I'm working with APIs that return cookies when logging in, and then I need to send those cookies back on every request. RapidAPI (formerly Paw) handles this beautifully.

**Reply by helloanoop** [reply]
@kevinrenskers Cookie support will be arriving soon. Added in the roadmap. Until then, please use this short script in your collection level scripts and you should be able to work with cookies https://github.com/usebruno/bruno/discussions/385#discussioncomment-7206789

**Reply by kevinrenskers** [reply]
Wow! This is really cool guys, I hadn't even seen the scripting support yet.

**Comment by majkinetor** [comment]
Any chance we have https://github.com/usebruno/bruno/issues/565 in the roadmap? Seems like a mandatory requirement for good behaving app which influences quite a lot of one's efficiency and joy of using bruno.

**Reply by helloanoop** [reply]
Agreed. Added as a priority on the roadmap for V2. We are planning a v1.0.0 launch on Oct 28th - https://github.com/usebruno/bruno/issues/643

**Comment by lllama** [comment]
Would be lovely to have #362 in 1.0. Running through an intercepting proxy is a strong requirement for our client work (both to do additional testing and for auditing).

**Comment by helloanoop** [comment]
We are set to launch [v1.0.0](https://github.com/usebruno/bruno/issues/643) on October 29, 2023, with our stable feature set. 

While I would have preferred to complete the OAuth2 and cookies features before announcing V1, the time left is too short. I also need to prepare for my [IndiaFoss 3.0 Talk & Project Showcase](https://indiafoss.net/2023). Unfortunately, I fell sick during the recent 4-day holiday in India, which dashed hopes of getting these across the goal line.

**v1.0.0** is not about having all the features but rather a message that our current platform is stable and suitable for use in production settings by teams. Many teams find our existing Bruno version to be perfectly sufficient for their needs.

It took 2 years to advance from **0.1.0** to **0.12.0**
And 1 month to go from **0.12.0** to **0.27.0** (since the rise that began on Sep 28th 2023)

We've been releasing updates every alternate day for the past month.
This incredible pace of development would not have been possible without the dedication, love, and contributions from our vibrant community. We've received valuable code contributions, identified and addressed issues, and gathered essential feedback.

Looking forward to Sun Oct 29th 2023.

Best,
Anoop

**Comment by rg-jaji** [comment]
Hi, I just discovered Bruno, awesome!
I will try to replace Postman in my company, one thing is missing to achieve this, I think.

It's the inheritance of auth (bearer token in my case) for the collection folder, it's very usefull in postman and this is my question, is it planned yet ? 

Thanks,
Best,
Romain

**Reply by Merdanpo** [reply]
Hello,
Collection tab lets you define Headers, Auth, Scripts, etc... for a collection. Requests in this collection will inherit these properties, it works like Postman do.

**Reply by zewa666** [reply]
same situation but I'm stuck with a Kerberos/NTLM environment, so having that as auth possibility would be huge

**Comment by dongokovacs** [comment]
it is the best api tool, we are waiting the performance possibility
after that i will push to change from postman to bruno gold :)

**Comment by bkserafi** [comment]
Kerberos/NTLM support will be great. For me it is the last thing that works in Postman, but does not in Bruno (Unless I do not know how to enable it.) Currently my credentials are not inherited from OS and authentication process is different in browser and in Bruno.
But great tool anyway!

**Comment by kritsadakornscb** [comment]
v. 2.0 , please support

- Tests>Snippets 
- Menu Import Curl (The same level as Import Collection)

**Comment by andreaippo** [comment]
Hello, any chance to get pretty-print for JSON responses?

Thanks a lot 🙏

**Reply by BlueRoseStory** [reply]
In addition to being "prettier", one nice feature would be to make any returned hyperlinks clickable.  Some APIs (Azure durable functions) return a bunch of links.   Being able to click those in the response would be sweet!  
btw, you have done an awesome job with this tool, and I am enjoying moving all my postman and thunderclient collections to Bruno.

**Comment by asgeirn** [comment]
My dev box is Fedora. Anything I can do to help get the RPM build up as a part of the release process?

**Comment by davidsebag0705** [comment]
Hi, I just discovered Bruno, awesome!

What about a skip request feature (see https://community.postman.com/t/the-new-skip-request-feature/54806) ?

Best,
David

**Comment by dheimoz** [comment]
hello, great idea, product and execution. I am a proud Golden user. I must thank and value the efforts, it is a great alternative to similar products, but with great data privacy focus. What is limiting my total use is the missing gRPC feature. I know it is on the roadmap, but do you have an estimate when would be released. Thanks again.

**Comment by pipl-rotemd** [comment]
Hi,  I am a new user. Love the privacy importance and how it is git-friendly.

I am really missing the possibility to comment out line/s in the request body (See [1451](https://github.com/usebruno/bruno/issues/1451)). 

Also, the ability to change the `Bearer Token` prefix,  e.g: instead of `Bearer` use `Token` ([django-rest-framework](https://www.django-rest-framework.org/api-guide/authentication/#tokenauthentication))

Thank you 🙏

**Comment by SharonWang000001** [comment]
Any support for adding binary as the request body in one POST request? 
I need to attcah one zip file as the request body, but now only find it support below,not having the binary.
<img width="722" alt="image" src="https://github.com/usebruno/bruno/assets/169419291/39432932-0dcb-4521-a209-2006210dc786">

**Reply by jeev20** [reply]
Hi @SharonWang000001 

I used the `Add File` under the `Multipart Form` option and could upload a file to my api endpoint. I am not sure if this is what you are looking for. 
![image](https://github.com/usebruno/bruno/assets/19647823/0361c7ca-aca1-404a-b57d-b1651282eaeb)

OR 

![image](https://github.com/usebruno/bruno/assets/19647823/c4214a20-c267-449d-88d6-d7c3557f71ad)

**Comment by beowulf79** [comment]
Environment can only be defined at Collection Level ? Any plan to move to a global level , easier to maintain and view ?

Thanks.

**Reply by BlueRoseStory** [reply]
My suggestion would be to keep them at the Collection level.  ThunderClient does the global level, and it is much more cluttered.  I have to create many environments there (they use different variables) and it is hard to tell one from another.   With Bruno having them at the collection level, I only see the environments that are specific to that collection.  For me, it is MUCH better.

**Comment by ThisIsBenny** [comment]
Hi, i'm not sure if this is the right place for Feature Request, but here is my request 😃 

I have compared a lot of Tools and i really like Bruno, but one thing which i more like on HTTPie UI is how they organize the Environment variables. Maybe it is possible to adapt it.

What HTTPie does is organize the environment variables in a table instead of different "sections," and you can define a default value.
This gives you a great overview of what is set in each environment, and you can use the default value if no value is specified for a specific environment. It ensures that I have the same variable across all environments.

At the moment i need to create in bruno a new variable on all environments manually. 

Example:
![image](https://github.com/user-attachments/assets/21725da4-5f62-4875-aeba-f024f2aa1ef6)

Thanks for the great work

**Reply by majkinetor** [reply]
I agree

In the meantime, you can basically automate this yourself as you have everything in environment folder in text files.
